### PR TITLE
Exposing image format information to LUA

### PR DESCRIPTION
**Pull request by npwoods**
Vas mentioned that there is major refactoring going on in the LUA engine code, so I suspect that this PR cannot be committed yet, but in any case I want it to get out for review.

### CODE REVIEW

File: src/frontend/mame/luaengine.cpp [hunk 1]
```diff
@@ -1771,6 +1772,35 @@ void lua_engine::initialize()
...
+	format_type.set("description", sol::property(&image_device_format::description));
+	format_type.set("optspec", sol::property(&image_device_format::optspec));
+	format_type.set("extensions", sol::property([this](const image_device_format &format) {
+			int index = 1;
+			sol::table option_table = sol().create_table();
+			for (const std::string &ext : format.extensions())
+				option_table[index++] = ext;
+			return option_table;
+		}));
+	sol().registry().set("format", format_type);
```
Comment (cracyc) [inline src/frontend/mame/luaengine.cpp:1803]: This isn't needed if you set the name above in the usertype declaration.

**Comment by npwoods** [comment]
Updated for recent changes

File: src/frontend/mame/luaengine.cpp [hunk 2]
```diff
@@ -1777,6 +1778,34 @@ void lua_engine::initialize()
...
+	image_type.set("formatlist", sol::property([this](const device_image_interface &image) {
+			sol::table format_table = sol().create_table();
+			for (const std::unique_ptr<image_device_format> &format : image.formatlist())
+				format_table[format->name()] = &*format;
+			return format_table;
+		}));
+
+/*	image_device_format library
+ *
+ * manager:machine().images[image_type].formatlist[x]
```
Comment (cuavas) [inline src/frontend/mame/luaengine.cpp:1790]: The image device is looked up by tag, and the format is looked up by name.

File: src/frontend/mame/luaengine.cpp [hunk 3]
```diff
@@ -1777,6 +1778,34 @@ void lua_engine::initialize()
...
+ * 
+ * format.name - name of the format (e.g. - "dsk")
+ * format.description - the description of the format
+ * format.extensions - all of the extensions, as an array
+ * format.optspec - the option spec associated with the format
+ * 
+ */
+	auto format_type = sol().registry().new_usertype<image_device_format>("format", "new", sol::no_constructor);
```
Comment (cuavas) [inline src/frontend/mame/luaengine.cpp:1798]: The name `format` is too generic.  You also don’t need `"new"` here.

**Review by cuavas — COMMENTED** [review]
This still doesn’t follow the same conventions as the other stuff for spacing.

Making tables for collection properties is convenient, but it’s expensive when there are potentially a lot of items, and you don’t get some of the convenience methods container wrappers get.  It’s probably not much of an issue for the image format extensions, but the image formats would probably benefit from being a proper container wrapper.

**Comment by npwoods** [comment]
I believe that I've addressed outstanding feedback.  A side effect of my change is that format list is now indexed by a one counted integer index, not the name.  It wasn't clear how to make the format name be the index while keeping the container object generic and not specific to image_device_format.

**Review by cuavas — COMMENTED** [review]
The `get` for this wrapper should do an O(n) lookup by name.  I’ll clean it up later along with some other stuff.  We’ll get this in for now, just be aware that the semantics of the image formats container will change.

**Pull request merged**